[MSE] Remove HTMLMedia Element access from MediaSource

jyavenard opened the pull request:
#### b0908a7024e145b4957d96953f1576ce62db91a8
<pre>
[MSE] Remove HTMLMedia Element access from MediaSource
<a href="https://bugs.webkit.org/show_bug.cgi?id=268796">https://bugs.webkit.org/show_bug.cgi?id=268796</a>
<a href="https://rdar.apple.com/122359957">rdar://122359957</a>

Reviewed by Eric Carlson.

If the MediaSource object lives in a worker thread, it will not be possible
for it to access the HTMLMediaElement directly.
We remove most access from the MediaSource to the media element.

One attribute accessed often is the HTMLMediaElement&apos;s currentTime value.
Instead we access the MediaPlayerPrivate value. As the HTMLMediaElement returns
the seekTime if a seek has been started we need a way to reproduce a similar
behaviour from the MediaPlayerPrivate. We add a `willSeekToTarget()` and `pendingSeekTime()`
method that will receive the seek time from the HTMLMediaElement.
To differentiate the behaviour between the actual player&apos;s time and what the
time according to the media element is, we add a `currentOrPendingSeekTime` method.

The handling of the Tracks hasn&apos;t been converted yet and will be done in
a follow-up change.

The MediaSourcePrivate::mediaTime method was duplicated in all implementations.
To simplify, we move it to the base class.

Fly-By: there were 3 virtual methods in the MediaPlayerPrivateInterface class
to retrieve the current time. We only leave on: currentMediaTime.
Fly-By: the static implementation to retrieve a SourceBuffer maximum size was
living in the MediaElementSession. We move it to the SourceBuffer instead.
Fly-By: The MockMediaPlayerMediaSource::isSeeking() would have always returned true
before seeking at least once. Correct logic.

Covered by existing tests, no change in observable behaviour.

* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::isOpen const): Deleted.
(WebCore::ManagedMediaSource::ensurePrefsRead):
(WebCore::ManagedMediaSource::monitorSourceBuffers):
(WebCore::ManagedMediaSource::isOpen const):
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::setPrivateAndOpen):
(WebCore::MediaSource::currentMediaTime const):
For consistency with other classes naming, rename currentTime() to currentMediaTime().
(WebCore::MediaSource::buffered const):
(WebCore::MediaSource::completeSeek):
(WebCore::MediaSource::hasBufferedTime):
(WebCore::MediaSource::hasCurrentTime):
(WebCore::MediaSource::hasFutureTime):
(WebCore::MediaSource::streamEndedWithError):
(WebCore::MediaSource::addSourceBuffer):
(WebCore::MediaSource::removeSourceBuffer):
(WebCore::MediaSource::detachFromElement):
(WebCore::MediaSource::attachToElement):
(WebCore::MediaSource::settings const):
(WebCore::MediaSource::stop):
(WebCore::MediaSource::notifyElementUpdateMediaState const):
(WebCore::MediaSource::incrementDroppedFrameCount):
(WebCore::MediaSource::addAudioTrackToElement):
(WebCore::MediaSource::addTextTrackToElement):
(WebCore::MediaSource::addVideoTrackToElement):
(WebCore::MediaSource::currentTime const): Deleted.
* Source/WebCore/Modules/mediasource/MediaSource.h:
(WebCore::MediaSource::mediaElement const): Deleted.
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::rangeRemoval):
(WebCore::SourceBuffer::changeType):
(WebCore::SourceBuffer::appendBufferInternal):
(WebCore::SourceBuffer::sourceBufferPrivateAppendComplete):
(WebCore::SourceBuffer::maximumBufferSize const):
(WebCore::SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment):
(WebCore::SourceBuffer::sourceBufferPrivateDidDropSample):
(WebCore::SourceBuffer::canPlayThroughRange):
(WebCore::SourceBuffer::settings const):
(WebCore::SourceBuffer::memoryPressure):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::deferredMediaSourceOpenCanProgress const):
move check to see if the ManagedMediaSource can progress to open to the HTMLMediaElement class.
(WebCore::HTMLMediaElement::seekWithTolerance):
(WebCore::HTMLMediaElement::seekTask):
(WebCore::HTMLMediaElement::clearSeeking):
(WebCore::HTMLMediaElement::maximumSourceBufferSize const): Deleted.
* Source/WebCore/html/HTMLMediaElement.h:
(WebCore::HTMLMediaElement::incrementDroppedFrameCount):
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::maximumMediaSourceBufferSize const): Deleted.
* Source/WebCore/html/MediaElementSession.h:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::willSeekToTarget):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.cpp:
(WebCore::MediaPlayerPrivateInterface::currentOrPendingSeekTime const):
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::currentMediaTime const):
(WebCore::MediaPlayerPrivateInterface::willSeekToTarget):
(WebCore::MediaPlayerPrivateInterface::pendingSeekTime const):
(WebCore::MediaPlayerPrivateInterface::currentTime const): Deleted.
(WebCore::MediaPlayerPrivateInterface::currentTimeDouble const): Deleted.
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
(WebCore::MediaSourcePrivate::currentMediaTime const):
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
To allow for a common currentMediaTime() implementation we need a generic
way to access the associated MediaPlayerPrivate.
Add a pure virtual `parent()`.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::currentTime const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::currentMediaTime const):
While seeking, return the last seekTime.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
(WebCore::MediaSourcePrivateAVFObjC::notifyActiveSourceBuffersChanged):
(WebCore::MediaSourcePrivateAVFObjC::player const):
(WebCore::MediaSourcePrivateAVFObjC::durationChanged):
(WebCore::MediaSourcePrivateAVFObjC::markEndOfStream):
(WebCore::MediaSourcePrivateAVFObjC::mediaPlayerReadyState const):
(WebCore::MediaSourcePrivateAVFObjC::setMediaPlayerReadyState):
(WebCore::MediaSourcePrivateAVFObjC::sourceBufferKeyNeeded):
(WebCore::MediaSourcePrivateAVFObjC::setSourceBufferWithSelectedVideo):
(WebCore::MediaSourcePrivateAVFObjC::currentMediaTime const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::player const):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.cpp:
(WebCore::MediaSourcePrivateGStreamer::currentMediaTime const): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.h:
* Source/WebCore/platform/graphics/win/MediaPlayerPrivateMediaFoundation.cpp:
Use MediaTime instead of floats.
(WebCore::MediaPlayerPrivateMediaFoundation::currentMediaTime const):
(WebCore::MediaPlayerPrivateMediaFoundation::currentTime const): Deleted.
* Source/WebCore/platform/graphics/win/MediaPlayerPrivateMediaFoundation.h:
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.cpp:
(WebCore::MockMediaPlayerMediaSource::seeking const): Fix logic, m_seekCompleted
was initialised to false, making seeking() return the wrong value.
(WebCore::MockMediaPlayerMediaSource::currentMediaTime const): Return last seek time
while seek hasn&apos;t completed.
(WebCore::MockMediaPlayerMediaSource::seekToTarget):
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.h:
* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.cpp:
(WebCore::MockMediaSourcePrivate::player const):
(WebCore::MockMediaSourcePrivate::currentMediaTime const): Deleted.
* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::willSeekToTarget):
(WebKit::MediaPlayerPrivateRemote::pendingSeekTime const):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::player const):
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/274442@main">https://commits.webkit.org/274442@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9eb28bb15ab384316bf4426e50132b0cc41f5504

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39126 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18081 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41479 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41660 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34843 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41432 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20955 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15431 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32747 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39700 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15231 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33914 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13227 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13193 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34851 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42937 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35527 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35180 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39013 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13938 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11499 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37243 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15544 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8751 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15207 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15030 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -1077,8 +1101,17 @@ bool MediaSource::virtualHasPendingActivity() const
 
+const Settings& MediaSource::settings() const
+{
+    ASSERT(scriptExecutionContext());
+    ASSERT(is<Document>(*scriptExecutionContext()));
```
Document isn't thread safe, so this should probably also assert that it is being called on the main thread

jyavenard replied:
indeed, it's a draft :)

at construction you can see that for now we assert that we are on the main thread, and this is where we define the m_dispatcher.

The aim later is to check what type of scriptExecutionContext we have and return the appropriate settings.

eric-carlson replied:
Sure, but in this version it is accessing Document so it should assert

eric-carlson commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -1220,6 +1233,13 @@ Document& SourceBuffer::document() const
 
+const Settings& SourceBuffer::settings() const
+{
+    ASSERT(scriptExecutionContext());
+    ASSERT(is<Document>(*scriptExecutionContext()));
+    return downcast<const Document>(*scriptExecutionContext()).settings();
```
Ditto

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -3571,6 +3573,7 @@ void HTMLMediaElement::seekTask()
     // 11 - Set the current playback position to the given new playback position
     m_seekRequested = true;
+    m_player->aboutToSeek(MediaTime::invalidTime());
     m_player->seekToTarget({ time, negativeTolerance, positiveTolerance });
```
Why not have `seekToTarget` clear it automatically?

jyavenard replied:
because seekToTarget has an implementation in all the MediaPlayerPrivate ; so I would have had to modify all of those; or make a default seekToTarget() implementation that calls like seekToTargetInternal() it wouldn't have been pretty.

eric-carlson replied:
Then do it in MediaPlayer?

jyavenard replied:
The MediaSourcePrivate only has access to the MediaPlayerPrivate

eric-carlson replied:
```
void MediaPlayer::seekToTarget(const SeekTarget& target)
{
    m_private->aboutToSeek(MediaTime::invalidTime());
    m_private->seekToTarget(target);
}
```

eric-carlson commented on Source/WebCore/platform/graphics/MediaPlayerPrivate.cpp:
```diff
@@ -62,6 +62,14 @@ auto MediaPlayerPrivateInterface::asyncVideoPlaybackQualityMetrics() -> Ref<Vide
+    if (aboutToSeekTime.isValid())
+        return aboutToSeekTime;
+    return currentMediaTime();
+}
```
It seems like this is just exposing an implementation detail. Why not have `MediaPlayerPrivateInterface:: currentMediaTime` do this automatically?

jyavenard replied:
I didn't want to introduce a regression, so this method is only called from the MediaSource, for the HTMLMediaElement it uses the same time type as before.

eric-carlson replied:
I don't have any great suggestions, but the name is confusing

eric-carlson replied:
It isn't great, but maybe `currentOrPendingSeekTime`?

eric-carlson commented on Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
```diff
@@ -125,15 +125,16 @@ class MediaPlayerPrivateInterface {
     virtual MediaTime getStartDate() const { return MediaTime::createWithDouble(std::numeric_limits<double>::quiet_NaN()); }
 
+    virtual void aboutToSeek(const MediaTime& time) { m_aboutToSeekTime = time; }
+    virtual MediaTime aboutToSeekTime() const { return m_aboutToSeekTime; }
```
Nit: `aboutToSeek` doesn't seem right. Maybe `willSeekToTarget` and `pendingSeekTime` instead?

jyavenard replied:
there were other methods named `aboutTo...` so I used that.

tbh, I changed the name of that method 3 times before this draft.

eric-carlson commented on Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
```diff
@@ -359,6 +360,7 @@ class MediaPlayerPrivateInterface {
 protected:
     mutable PlatformTimeRanges m_seekable;
     bool m_shouldCheckHardwareSupport { false };
+    MediaTime m_aboutToSeekTime { MediaTime::invalidTime() };
```
Nit: Ditto, maybe `m_pendingSeekTime`?

eric-carlson commented on Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
```diff
@@ -339,6 +339,18 @@ MediaTime MediaPlayerPrivateRemote::currentMediaTime() const
+{
+    Locker locker { m_currentTimeEstimator.lock() };
+    return MediaPlayerPrivateInterface::aboutToSeekTime();
+}
```
Ditto

eric-carlson commented on Source/WebCore/platform/graphics/MediaPlayerPrivate.cpp:
```diff
@@ -62,6 +62,14 @@ auto MediaPlayerPrivateInterface::asyncVideoPlaybackQualityMetrics() -> Ref<Vide
+    if (pendingSeekTime.isValid())
+        return pendingSeekTime;
+    return currentMediaTime();
+}
```
Because this is only used for MediaSource, why not make `pendingSeekTime ` public and put the logic in `MediaSourcePrivate::currentMediaTime`

jyavenard replied:
I need the access to m_pendingSeeking to be thread-safe and atomic with the currentMediaTime, at this stage thread-safety is an implementation details and so is better managed by the MediaPlayerPrivate.

thinking about atomicity, the current code is wrong, I need to make that method virtual..

eric-carlson commented on Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
```diff
@@ -189,6 +189,13 @@ void MediaSourcePrivate::ensureOnDispatcher(Function<void()>&& function) const
+{
+    if (RefPtr player = this->player())
+        return player->apparentMediaTime();
+    return MediaTime::invalidTime();
```
```suggestion
    RefPtr player = this->player();
    if (!player)
        return MediaTime::invalidTime();

    auto pendingSeekTime = player->pendingSeekTime();
    return pendingSeekTime.isValid() ? pendingSeekTime : player-> currentMediaTime();
```

jyavenard replied:
issue with this is that it is possible for pendingSeekTime to change before we call currentMediaTime.

Likelihood is going to be very low, but it is necessary I believe for having synchronised read of both pendingSeekTime and currentMediaTime.

jyavenard committed:
[MSE] Remove HTMLMedia Element access from MediaSource
https://bugs.webkit.org/show_bug.cgi?id=268796
rdar://122359957

Reviewed by Eric Carlson.

If the MediaSource object lives in a worker thread, it will not be possible
for it to access the HTMLMediaElement directly.
We remove most access from the MediaSource to the media element.

One attribute accessed often is the HTMLMediaElement's currentTime value.
Instead we access the MediaPlayerPrivate value. As the HTMLMediaElement returns
the seekTime if a seek has been started we need a way to reproduce a similar
behaviour from the MediaPlayerPrivate. We add a `willSeekToTarget()` and `pendingSeekTime()`
method that will receive the seek time from the HTMLMediaElement.
To differentiate the behaviour between the actual player's time and what the
time according to the media element is, we add a `currentOrPendingSeekTime` method.

The handling of the Tracks hasn't been converted yet and will be done in
a follow-up change.

The MediaSourcePrivate::mediaTime method was duplicated in all implementations.
To simplify, we move it to the base class.

Fly-By: there were 3 virtual methods in the MediaPlayerPrivateInterface class
to retrieve the current time. We only leave on: currentMediaTime.
Fly-By: the static implementation to retrieve a SourceBuffer maximum size was
living in the MediaElementSession. We move it to the SourceBuffer instead.
Fly-By: The MockMediaPlayerMediaSource::isSeeking() would have always returned true
before seeking at least once. Correct logic.

Covered by existing tests, no change in observable behaviour.

* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::isOpen const): Deleted.
(WebCore::ManagedMediaSource::ensurePrefsRead):
(WebCore::ManagedMediaSource::monitorSourceBuffers):
(WebCore::ManagedMediaSource::isOpen const):
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::setPrivateAndOpen):
(WebCore::MediaSource::currentMediaTime const):
For consistency with other classes naming, rename currentTime() to currentMediaTime().
(WebCore::MediaSource::buffered const):
(WebCore::MediaSource::completeSeek):
(WebCore::MediaSource::hasBufferedTime):
(WebCore::MediaSource::hasCurrentTime):
(WebCore::MediaSource::hasFutureTime):
(WebCore::MediaSource::streamEndedWithError):
(WebCore::MediaSource::addSourceBuffer):
(WebCore::MediaSource::removeSourceBuffer):
(WebCore::MediaSource::detachFromElement):
(WebCore::MediaSource::attachToElement):
(WebCore::MediaSource::settings const):
(WebCore::MediaSource::stop):
(WebCore::MediaSource::notifyElementUpdateMediaState const):
(WebCore::MediaSource::incrementDroppedFrameCount):
(WebCore::MediaSource::addAudioTrackToElement):
(WebCore::MediaSource::addTextTrackToElement):
(WebCore::MediaSource::addVideoTrackToElement):
(WebCore::MediaSource::currentTime const): Deleted.
* Source/WebCore/Modules/mediasource/MediaSource.h:
(WebCore::MediaSource::mediaElement const): Deleted.
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::rangeRemoval):
(WebCore::SourceBuffer::changeType):
(WebCore::SourceBuffer::appendBufferInternal):
(WebCore::SourceBuffer::sourceBufferPrivateAppendComplete):
(WebCore::SourceBuffer::maximumBufferSize const):
(WebCore::SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment):
(WebCore::SourceBuffer::sourceBufferPrivateDidDropSample):
(WebCore::SourceBuffer::canPlayThroughRange):
(WebCore::SourceBuffer::settings const):
(WebCore::SourceBuffer::memoryPressure):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::deferredMediaSourceOpenCanProgress const):
move check to see if the ManagedMediaSource can progress to open to the HTMLMediaElement class.
(WebCore::HTMLMediaElement::seekWithTolerance):
(WebCore::HTMLMediaElement::seekTask):
(WebCore::HTMLMediaElement::clearSeeking):
(WebCore::HTMLMediaElement::maximumSourceBufferSize const): Deleted.
* Source/WebCore/html/HTMLMediaElement.h:
(WebCore::HTMLMediaElement::incrementDroppedFrameCount):
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::maximumMediaSourceBufferSize const): Deleted.
* Source/WebCore/html/MediaElementSession.h:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::willSeekToTarget):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.cpp:
(WebCore::MediaPlayerPrivateInterface::currentOrPendingSeekTime const):
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::currentMediaTime const):
(WebCore::MediaPlayerPrivateInterface::willSeekToTarget):
(WebCore::MediaPlayerPrivateInterface::pendingSeekTime const):
(WebCore::MediaPlayerPrivateInterface::currentTime const): Deleted.
(WebCore::MediaPlayerPrivateInterface::currentTimeDouble const): Deleted.
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
(WebCore::MediaSourcePrivate::currentMediaTime const):
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
To allow for a common currentMediaTime() implementation we need a generic
way to access the associated MediaPlayerPrivate.
Add a pure virtual `parent()`.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::currentTime const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::currentMediaTime const):
While seeking, return the last seekTime.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
(WebCore::MediaSourcePrivateAVFObjC::notifyActiveSourceBuffersChanged):
(WebCore::MediaSourcePrivateAVFObjC::player const):
(WebCore::MediaSourcePrivateAVFObjC::durationChanged):
(WebCore::MediaSourcePrivateAVFObjC::markEndOfStream):
(WebCore::MediaSourcePrivateAVFObjC::mediaPlayerReadyState const):
(WebCore::MediaSourcePrivateAVFObjC::setMediaPlayerReadyState):
(WebCore::MediaSourcePrivateAVFObjC::sourceBufferKeyNeeded):
(WebCore::MediaSourcePrivateAVFObjC::setSourceBufferWithSelectedVideo):
(WebCore::MediaSourcePrivateAVFObjC::currentMediaTime const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::player const):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.cpp:
(WebCore::MediaSourcePrivateGStreamer::currentMediaTime const): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.h:
* Source/WebCore/platform/graphics/win/MediaPlayerPrivateMediaFoundation.cpp:
Use MediaTime instead of floats.
(WebCore::MediaPlayerPrivateMediaFoundation::currentMediaTime const):
(WebCore::MediaPlayerPrivateMediaFoundation::currentTime const): Deleted.
* Source/WebCore/platform/graphics/win/MediaPlayerPrivateMediaFoundation.h:
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.cpp:
(WebCore::MockMediaPlayerMediaSource::seeking const): Fix logic, m_seekCompleted
was initialised to false, making seeking() return the wrong value.
(WebCore::MockMediaPlayerMediaSource::currentMediaTime const): Return last seek time
while seek hasn't completed.
(WebCore::MockMediaPlayerMediaSource::seekToTarget):
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.h:
* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.cpp:
(WebCore::MockMediaSourcePrivate::player const):
(WebCore::MockMediaSourcePrivate::currentMediaTime const): Deleted.
* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::willSeekToTarget):
(WebKit::MediaPlayerPrivateRemote::pendingSeekTime const):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::player const):
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.h:

Canonical link: https://commits.webkit.org/274442@main
